AppCenterDataStorage/AppCenterDataStorage/Internal/Client/MSTokenExchange.m

```diff
@@ -1,45 +1,162 @@
+            return;
           }
 
           // Create token result object.
-          MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithDictionary:jsonDictionary];
+          MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithDictionary:jsonDictionary[kMSTokens][0]];
+
+          // Token exchange failed to give back a token
+          if (error) {
+            MSLogError([MSDataStore logTag], @"Get on DB Token had an error with code: %td, description: %@", error.code,
+                       error.localizedDescription);
+            completion([[MSTokensResponse alloc] initWithTokens:nil], error);
+            return;
+          }
+
+          // Create token response object.
+          MSTokensResponse *tokens = [[MSTokensResponse alloc] initWithTokens:@[ tokenResult ]];
+
+          // Token exchange did not get back an error but aquiring the token did not succeed either
+          if (tokenResult && tokenResult.status != kMSTokenResultSucceed) {
+            MSLogError([MSDataStore logTag], @"Token result had a status of %@", tokenResult.status);
+            completion(tokens, error);
+            return;
+          }
+
+          // Cache the newly acquired token.
+          [MSTokenExchange saveToken:tokenResult];
           completion(tokens, error);
-        }
-      }];
+        }];
+  } else {
+    completion([[MSTokensResponse alloc] initWithTokens:@[ cachedToken ]], nil);
+  }
+}
+
+/*
+ * Cache the Cosmos DB token received from the token exchange service.
+ * The token is stored in KeyChain
+ */
++ (void)saveToken:(MSTokenResult *)tokenResult {
+  if (tokenResult) {
+    NSString *tokenString = [tokenResult serializeToString];
+    BOOL success = [MSKeychainUtil storeString:tokenString forKey:[MSTokenExchange tokenKeyNameForPartition:tokenResult.partition]];
+    if (success) {
+      MSLogDebug([MSDataStore logTag], @"Saved token in keychain for the partitionKey : %@.", tokenResult.partition);
+    } else {
+      MSLogError([MSDataStore logTag], @"Failed to save the token in keychain for the partitionKey : %@.", tokenResult.partition);
+    }
+  }
+}
+
+/*
+ * Get back the Cached Cosmos DB token
+ * If the DB token has expired, that token is deleted from KeyChain
+ */
++ (MSTokenResult *)retrieveCachedToken:(NSString *)partitionName {
+  if (partitionName) {
+    NSString *tokenString = [MSKeychainUtil stringForKey:[MSTokenExchange tokenKeyNameForPartition:partitionName]];
+    if (tokenString) {
+      MSTokenResult *tokenResult = [[MSTokenResult alloc] initWithString:tokenString];
+      NSDate *currentUTCDate = [NSDate date];
+      NSDate *tokenExpireDate = [MSUtility dateFromISO8601:tokenResult.expiresOn];
+      if ([currentUTCDate laterDate:tokenExpireDate] == currentUTCDate) {
+        [MSTokenExchange removeCachedToken:partitionName];
+        MSLogWarning([MSDataStore logTag], @"The token in the cache has expired for the partitionKey : %@.", partitionName);
```

Comment:
nit: it would be better to log a message about deleting token from cache.

```The token in the cache for the partitionKey ... has expired hence being removed from cache.```

Reply:
Ignore my previous comment. It's already logged in `removeCachedToken`.